pkg/registry/core/service/storage/rest.go

```diff
@@ -696,51 +678,102 @@ func initClusterIP(service *api.Service, allocator ipallocator.Interface) (bool,
-				if err != nil {
-					// TODO: when validation becomes versioned, this gets more complicated.
-					el := field.ErrorList{field.Invalid(field.NewPath("spec", "ports").Index(i).Child("nodePort"), np, err.Error())}
-					return errors.NewInvalid(api.Kind("Service"), service.Name, el)
-				}
-				servicePort.NodePort = int32(np)
-				svcPortToNodePort[int(servicePort.Port)] = np
-			} else {
-				nodePort, err := nodePortOp.AllocateNext()
-				if err != nil {
-					// TODO: what error should be returned here?  It's not a
-					// field-level validation failure (the field is valid), and it's
-					// not really an internal error.
-					return errors.NewInternalError(fmt.Errorf("failed to allocate a nodePort: %v", err))
-				}
-				servicePort.NodePort = int32(nodePort)
-				svcPortToNodePort[int(servicePort.Port)] = nodePort
-			}
-		} else if int(servicePort.NodePort) != allocatedNodePort {
-			// TODO(xiangpengzhao): do we need to allocate a new NodePort in this case?
-			// Note: the current implementation is better, because it saves a NodePort.
-			if servicePort.NodePort == 0 {
-				servicePort.NodePort = int32(allocatedNodePort)
-			} else {
-				err := nodePortOp.Allocate(int(servicePort.NodePort))
-				if err != nil {
-					// TODO: when validation becomes versioned, this gets more complicated.
-					el := field.ErrorList{field.Invalid(field.NewPath("spec", "ports").Index(i).Child("nodePort"), servicePort.NodePort, err.Error())}
-					return errors.NewInvalid(api.Kind("Service"), service.Name, el)
-				}
+		nodePort, visited := specifiedNodePorts[servicePort.Port]
+		if visited && servicePort.NodePort == 0 {
+			servicePort.NodePort = nodePort
+		}
+	}
+
+	// === Allocate explicitly specified NodePorts
+	svcPortToAllocatedNodePort := map[int32]int32{}
+	for i, servicePort := range service.Spec.Ports {
+		allocatedNodePort, visited := svcPortToAllocatedNodePort[servicePort.Port]
+		// Same port with different protocol
+		if visited && allocatedNodePort == servicePort.NodePort {
+			continue
+		}
+		if servicePort.NodePort > 0 {
+			err := nodePortOp.Allocate(int(servicePort.NodePort))
+			if err != nil {
+				el := field.ErrorList{field.Invalid(
+					specPath.Child("ports").Index(i).Child("nodePort"),
+					servicePort.NodePort,
+					err.Error())}
+				return errors.NewInvalid(api.Kind("Service"), service.Name, el)
 			}
+			svcPortToAllocatedNodePort[servicePort.Port] = servicePort.NodePort
 		}
 	}
 
+	// Phase 2: allocate remaining NodePorts
+	if needHealthCheck && healthPort == 0 {
+		err := allocateHealthCheckNodePort(service, nodePortOp)
+		if err != nil {
+			// TODO: what error should be returned here?  It's not a
+			// field-level validation failure (the field is valid), and it's
+			// not really an internal error.
+			return errors.NewInternalError(err)
+		}
+	}
+	for i := range service.Spec.Ports {
+		servicePort := &service.Spec.Ports[i]
+		if servicePort.NodePort == 0 {
+			nodePort, err := nodePortOp.AllocateNext()
```

Comment:
I think that the only error possible here is that there are no ports free in the allocator, so it NewInternalError may make sense

Reply:
Good to know that! I've removed that comment.